python/http_ece/__init__.py

```diff
@@ -65,39 +70,41 @@ def derive_key(mode, version, salt=None, key=None, dh=None, auth_secret=None,
     def build_info(base, info_context):
         return b"Content-Encoding: " + base + b"\0" + info_context
 
-    def derive_dh(mode, version, dh, keyid, keymap, keylabels):
+    def derive_dh(mode, version, private_key, dh, label):
         def length_prefix(key):
             return struct.pack("!H", len(key)) + key
 
-        if keyid is None:
-            raise ECEException(u"'keyid' is not specified with 'dh'")
-        if keyid not in keymap:
-            raise ECEException(u"'keyid' doesn't identify a key: " + keyid)
         if mode == "encrypt":
-            sender_pub_key = key or keymap[keyid].get_pubkey()
+            sender_pub_key = private_key.get_pubkey()
             receiver_pub_key = dh
         else:
             sender_pub_key = dh
-            receiver_pub_key = key or keymap[keyid].get_pubkey()
+            receiver_pub_key = private_key.get_pubkey()
+
         if version == "aes128gcm":
             context = b"WebPush: info\x00" + receiver_pub_key + sender_pub_key
         else:
-            label = keylabels.get(keyid, 'P-256').encode('utf-8')
             context = (label + b"\0" + length_prefix(receiver_pub_key) +
                        length_prefix(sender_pub_key))
 
-        return keymap[keyid].get_ecdh_key(dh), context
+        return private_key.get_ecdh_key(dh), context
 
     if version not in versions:
         raise ECEException(u"Invalid version")
     if mode not in ['encrypt', 'decrypt']:
         raise ECEException(u"unknown 'mode' specified: " + mode)
-    if salt is None or len(salt) != 16:
+    if salt is None or len(salt) != KEY_LENGTH:
         raise ECEException(u"'salt' must be a 16 octet value")
-    if dh is not None:
-        (secret, context) = derive_dh(mode=mode, version=version, dh=dh,
-                                      keyid=keyid, keymap=keymap,
-                                      keylabels=keylabels)
+    if dh is not None or private_key is not None:
+        # We need a key in the keymap unless we're decrypting 'aes128gcm', where
+        # we can get the key from the keyid.
+        if private_key is None:
+            raise ECEException(u"DH requires a private_key")
+        label = keylabels.get(keyid, 'P-256').encode('utf-8')
+
+        (secret, context) = derive_dh(mode=mode, version=version,
+                                      private_key=private_key, dh=dh,
+                                      label=label)
     elif keyid in keymap:
         secret = keymap[keyid]
```

Comment:
Are we dropping `keymap` in favor of passing the key directly? (I'm a fan of doing so, but we should be clear across the board about it.)